### PR TITLE
Merge native programs parts into one unit

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

Programs are artificially split into an `api` and a `program` crate.  The motivating cause of this is when one program depends directly on another the other's entrypoint might be exposed instead of the primary program's.  Also, linking on Linux causes duplicate symbol errors for the entrypoint symbol `process`

#### Summary of Changes

This PR starts with `config`, `stake`, `vest`, and `vote` since they are the more complicated cases, good representations of the problem, and depended on both by other crates but also by other programs.

- Define a new macro that consolidates the program definition
- Use unique entrypoints for each module that is the same as the module "name"
- Combine each program into a single unit.  
For example, `config_ap/program` becomes:
package = "solana-config-program" located in `/programs/config` and produces `solana_config_program.rlib` and `solana_config_program.dylib`.

Fixes #

### CODE REVIEW

File: genesis-programs/src/lib.rs [hunk 1]
```diff
@@ -81,7 +81,7 @@ pub fn get_programs(operating_mode: OperatingMode, epoch: Epoch) -> Option<Vec<(
                 // The epoch of std::u64::MAX - 1 is a placeholder and is expected to be reduced in
                 // a future hard fork.
                 Some(vec![
-                    solana_config_program!(),
+                    solana_config_program::solana_config_program!(),
```
Comment (mvines) [inline genesis-programs/src/lib.rs]: This looks a little funny: `solana_config_program::solana_config_program!()`

File: programs/stake/src/lib.rs [hunk 2]
```diff
@@ -12,9 +13,11 @@ const STAKE_PROGRAM_ID: [u8; 32] = [
     120, 114, 43, 104, 164, 157, 192, 0, 0, 0, 0,
 ];
 
-solana_sdk::solana_name_id!(
+solana_sdk::solana_program!(
```
Comment (garious) [inline programs/stake/src/lib.rs]: This macro can use a better name. Maybe `declare_program!()`

File: sdk/src/instruction_processor_utils.rs [hunk 3]
```diff
@@ -27,6 +27,29 @@ macro_rules! solana_entrypoint(
...
+
+        #[macro_export]
+        macro_rules! $name {
+            () => {
+                (stringify!($name).to_string(), $crate::id())
+            };
+        }
+
+        #[no_mangle]
+        pub extern "C" fn $name(
```
Comment (garious) [inline sdk/src/instruction_processor_utils.rs]: This should probably include a suffix like `_entrypoint`. Then `$name` can be just one word, such as `stake` or `config`.

File: sdk/src/instruction_processor_utils.rs [hunk 4]
```diff
@@ -27,6 +27,29 @@ macro_rules! solana_entrypoint(
     )
 );
 
+#[macro_export]
+macro_rules! solana_program(
+    ($id:ident, $bs58:expr, $name:ident, $entrypoint:ident) => (
```
Comment (garious) [inline sdk/src/instruction_processor_utils.rs]: Need to document these parameters in detail. Even after looking at the implementation, I can quite figure out the purpose of each. And hopefully adding docs will expose some better parameter names.

File: genesis-programs/src/lib.rs (hunk 1 above)
Comment (jackcmay) [reply genesis-programs/src/lib.rs]: Yup, that is one of two changes I have in the wings, didn't want to kick ci for a couple of liittle nits.

File: programs/stake/src/lib.rs (hunk 2 above)
Comment (jackcmay) [reply programs/stake/src/lib.rs]: Should we also change the `solana_name_id` macro's name as well?

File: sdk/src/instruction_processor_utils.rs (hunk 3 above)
Comment (jackcmay) [reply sdk/src/instruction_processor_utils.rs]: `name` is also used for the filename, we might want to keep the long name like `solana_stake_program`.  The particular naming of the entrypoint is pretty much invisible to everyone but the native loader.  Using the package name would be ideal.

Comment (jackcmay) [reply sdk/src/instruction_processor_utils.rs]: Could use:
```
const NAME: &'static str = env!("CARGO_PKG_NAME");
```

But would have to lower dashes to underlines

**Pull request merged**